Search result crashes

When trying to perform a search and pressing on a result, I get this crash:
> 2018-08-21 12:12:58.411 21463-21463/io.plaidapp E/AndroidRuntime: FATAL EXCEPTION: main
    Process: io.plaidapp, PID: 21463
    java.lang.RuntimeException: Unable to start activity ComponentInfo{io.plaidapp/io.plaidapp.dribbble.ui.shot.ShotActivity}: java.lang.IllegalStateException: Could not retrieve shot 4635899
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2913)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3048)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1808)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
     Caused by: java.lang.IllegalStateException: Could not retrieve shot 4635899
        at io.plaidapp.dribbble.ui.shot.ShotViewModel.<init>(ShotViewModel.kt:50)
        at io.plaidapp.dribbble.ui.shot.ShotViewModelFactory.create(ShotViewModelFactory.kt:40)
        at android.arch.lifecycle.ViewModelProvider.get(ViewModelProvider.java:134)
        at android.arch.lifecycle.ViewModelProvider.get(ViewModelProvider.java:102)
        at io.plaidapp.dribbble.ui.shot.ShotActivity.onCreate(ShotActivity.kt:126)
        at android.app.Activity.performCreate(Activity.java:7136)
        at android.app.Activity.performCreate(Activity.java:7127)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1271)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2893)

--- a/core/src/main/java/io/plaidapp/core/data/SearchDataManager.java
+++ b/core/src/main/java/io/plaidapp/core/data/SearchDataManager.java
@@ -18,20 +18,16 @@
 package io.plaidapp.core.data;
 
 import android.content.Context;
-
+import io.plaidapp.core.designernews.Injection;
 import io.plaidapp.core.designernews.domain.SearchStoriesUseCase;
+import io.plaidapp.core.dribbble.DribbbleInjection;
+import io.plaidapp.core.dribbble.data.ShotsRepository;
+import io.plaidapp.core.dribbble.data.api.model.Shot;
+import kotlin.Unit;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
-import java.util.ArrayList;
 import java.util.List;
-
-import io.plaidapp.core.dribbble.data.search.DribbbleSearchService;
-import io.plaidapp.core.dribbble.data.api.model.Shot;
-import io.plaidapp.core.designernews.Injection;
-import retrofit2.Call;
-import retrofit2.Callback;
-import retrofit2.Response;
 
 /**
  * Responsible for loading search results from dribbble and designer news. Instantiating classes are
@@ -41,15 +37,16 @@ public abstract class SearchDataManager extends BaseDataManager<List<? extends P
         implements LoadSourceCallback {
 
     private final SearchStoriesUseCase searchStoriesUseCase;
+    private final ShotsRepository shotsRepository;
+
     // state
     private String query = "";
     private int page = 1;
-    private List<Call> inflight;
 
     public SearchDataManager(Context context) {
         super();
         searchStoriesUseCase = Injection.provideSearchStoriesUseCase(context);
-        inflight = new ArrayList<>();
+        shotsRepository = DribbbleInjection.provideShotsRepository();
     }
 
     public void searchFor(String query) {
@@ -76,13 +73,8 @@ public abstract class SearchDataManager extends BaseDataManager<List<? extends P
 
     @Override
     public void cancelLoading() {
-        if (inflight.size() > 0) {
-            for (Call call : inflight) {
-                call.cancel();
-            }
-            inflight.clear();
-            searchStoriesUseCase.cancelAllRequests();
-        }
+        searchStoriesUseCase.cancelAllRequests();
+        shotsRepository.cancelAllSearches();
     }
 
     public String getQuery() {
@@ -97,38 +89,23 @@ public abstract class SearchDataManager extends BaseDataManager<List<? extends P
 
     private void searchDribbble(final String query, final int resultsPage) {
         loadStarted();
-        final Call<List<Shot>> dribbbleSearchCall = getDribbbleSearchApi().search(
-                query, resultsPage, DribbbleSearchService.PER_PAGE_DEFAULT,
-                DribbbleSearchService.SORT_POPULAR);
-        dribbbleSearchCall.enqueue(new Callback<List<Shot>>() {
-            @Override
-            public void onResponse(Call<List<Shot>> call, Response<List<Shot>> response) {
-                if (response.isSuccessful()) {
-                    loadFinished();
-                    final List<Shot> shots = response.body();
-                    if (shots != null) {
-                        setPage(shots, resultsPage);
-                        setDataSource(shots,
-                                Source.DribbbleSearchSource.DRIBBBLE_QUERY_PREFIX + query);
-                        onDataLoaded(shots);
-                    }
-                    inflight.remove(dribbbleSearchCall);
-                } else {
-                    failure(dribbbleSearchCall);
-                }
+        shotsRepository.search(query, page, result -> {
+            loadFinished();
+            if (result instanceof Result.Success) {
+                List<Shot> shots = ((Result.Success<? extends List<Shot>>) result).getData();
+                setPage(shots, resultsPage);
+                setDataSource(shots,
+                        Source.DribbbleSearchSource.DRIBBBLE_QUERY_PREFIX + query);
+                onDataLoaded(shots);
             }
-
-            @Override
-            public void onFailure(Call<List<Shot>> call, Throwable t) {
-                failure(dribbbleSearchCall);
-            }
+            return Unit.INSTANCE;
         });
-        inflight.add(dribbbleSearchCall);
+
     }
 
     @Override
     public void sourceLoaded(@Nullable List<? extends PlaidItem> result, int page,
-            @NotNull String source) {
+                             @NotNull String source) {
         loadFinished();
         if (result != null) {
             setPage(result, page);
@@ -141,11 +118,6 @@ public abstract class SearchDataManager extends BaseDataManager<List<? extends P
     @Override
     public void loadFailed(@NotNull String source) {
         loadFinished();
-    }
-
-    private void failure(Call call) {
-        loadFinished();
-        inflight.remove(call);
     }
 
 }

